src/Oden/Compiler/Validation/Typed.hs

```diff
@@ -108,18 +112,33 @@ validateExpr = void . traverseExpr identityTraversal { onExpr = onExpr'
       Unresolved protocol method _ -> protocol == divisionName && method == divideName
       Resolved protocol method _ -> protocol == divisionName && method == divideName
 
+
+validateMainPkg :: [TypedDefinition] -> Validate ()
+validateMainPkg defs = case find isMainFn defs of
+    Just (Definition _ _ (_, NoArgFn _ _ (TNoArgFn _ (TCon _ (FQN (NativePackageName []) (Identifier "unit")))))) -> return ()
+    Just def -> throwError . MainPkgDoesNotHaveValidMainFn . definitionSourceInfo $ def
+    Nothing -> throwError MainPkgDoesNotHaveMainFn
```

Comment:
Perhaps we can have a `SourceInfo` value on `MainPkgDoesNotHaveMainFn` as well? The package declaration's SourceInfo would be the best, I think.